Ray for federated learning and privacy-preserving computing (#17)
ray-project/enhancements: reps/2022-12-08-ray-for-federated-learning-and-privacy-preserving-computing.md

### General Motivation
Federated machine learning and privacy-preserving computing are getting wider. They're obviously distributed systems because they're always across parties(companies or ORGs.) To extend Ray ecosystem to federated learning and privacy-preserving computing domain, we should let users be easy to use Ray, for building their own federated learning and privacy-preserving computing applications without any concern on data privacy, task-attacks, illegal invasion, and etc.

PS: I also found another related issue: https://github.com/ray-project/ray/issues/25846

In this proposal, we'd like to build a connector layer on Ray, to provide the ability for users use Ray easily to build their such kind of system, with the secure issues addressed totally.

### Key requirements:
- Provide the ability for users to easily build their such kind of application on Ray.
- Setup different Ray clusters for different parties to avoid uncotrollable complex Ray protocols.
- Have a unified and global-viewed programming mode for different parties.
- Data transmition across parties should be in push mode instead of pull mode.

> - Tasks should be driven in multi-controller mode. That means tasks should be driven by themselves(who are inside this party) instead of others.

Author: details are supplemented.

### Should this change be within ray or outside?
Yes.

> 1. A `ray/python/ray/fed` directory should be added to ray repo as the connector layer, like ray collective, ray dag.

Reviewer: For not introducing unnecessary heavy dependencies, I believe it's better to create a new repository instead of introducing it into `ray` repo.

So the current section should be:
```md
### Should this change be within ray or outside?
No
1. Adding a new repo `RayFed` under `ray-project`: `ray-project/RayFed`
2. The package name is `rayfed`: pip install rayfed
3. importing lines should be `import fed`
```

Author: OK, it can be an independent repo for now.

2. `from ray import fed` this kind of code should be supported in Ray core, and then there is no more requirement that needs to change Ray core code yet.

## Stewardship
### Required Reviewers
@jovany-wang
### Shepherd of the Proposal (should be a senior committer)

## Design and Architecture
The key words in this design are `multiple controller mode`, `restricted data perimeters`.

### User Examples
#### A Simple Example
```python

> from ray import fed

Reviewer: ```suggestion
import fed
```

### Benefits 
Compared with running the DAG in one Ray cluster, we have the following significant benefits:
- It's very close to Ray native programming pattern. But the DAG could be ran across Ray clusters.
- Very restricted and clear data perimeters. we only transmit the data which is needed by others. And all of the data couldn't be fetched in any way if we don't allow.
- If we run the DAG in one Ray cluster, data transmition is in PULL-BASED mode. For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.

> - All the disadvanteges are addressed in this proposal: code distribution, data privacy, task-attacks, illegal invasion, deserialization vulnerability，and etc.

Reviewer: We can highlight these disadvantages.

> - All the disadvanteges are addressed in this proposal: code distribution, data privacy, task-attacks, illegal invasion, deserialization vulnerability，and etc.

Reviewer: ```suggestion
- All the disadvantages are addressed in this proposal: code distribution, data privacy, task-attacks, illegal invasion, deserialization vulnerability，and etc.
```

- Brings all Ray advantages to the local Ray cluster, like highly performance RPC, fault tolerance, task scheduling/resource management, and other Ray ecosystem libraries（ Ray datasets, Ray train and Ray serve） for local training.

## Compatibility, Deprecation, and Migration Plan
None.

## Test Plan and Acceptance Criteria

> - Test should includes unit tests and integration tests(release tests).

Reviewer: ```suggestion
- Test should include unit tests and integration tests(release tests).
```

Author: Nice catch

- Tests should be enabled in CI.
- Benchmark on across parties data transmition for typical federated learning case.
- Document page and necessary code comments.

## (Optional) Follow-on Work
- UX improvements for across parties exceptions passing.
- Provide single-controller mode for quickly building applications and debug inner one party.

> - Performance optimization for other workloads.

Reviewer: Maybe 2 more items should be added in this section:
- A unified gateway for ETH communication.
- Support TEE(Trust Execution Environment) device.

Author: Yea, nice suggestions.

- Have a unified and global-viewed programming mode for different parties.
- Data transmission across parties should be in push mode instead of pull mode. It's hard to prevent malicious attacker stealing data in pull mode. Push mode is much better since it's the responsibility of the party to decide whether to send data to others.
- Tasks should be driven in multi-controller mode. That means tasks should be driven by themselves(who are inside this party) instead of others.

### Should this change be within ray or outside?
No, there is no need to change ray repo, the steps should be:
1. Adding a new repo `RayFed` under `ray-project`: `ray-project/RayFed`
2. The package name is `rayfed`: pip install rayfed
3. importing lines should be `import fed`

## Stewardship
### Required Reviewers
@jovany-wang
### Shepherd of the Proposal (should be a senior committer)

> 

Reviewer: @jovany-wang

> 

Reviewer: ```suggestion
@jovany-wang 
```

PS: I also found another related issue: https://github.com/ray-project/ray/issues/25846

In this proposal, we'd like to build a connector layer on Ray, to provide the ability for users use Ray easily to build their such kind of system, with the secure issues addressed totally.

### Key requirements:
- Provide the ability for users to easily build their such kind of application on Ray.
- Setup different Ray clusters for different parties to avoid uncotrollable complex Ray protocols. 
  - We have tried to setup a single cross-party Ray cluster but failed for security reasons. It is difficult for parties to detect and prevent attacks, e.g. a malicious attacker (it could even be one of the parties) runs destructive code. The complex Ray protocols make it very difficult to enhance the security under single Ray cluster. 
- Have a unified and global-viewed programming mode for different parties.

> - Data transmition across parties should be in push mode instead of pull mode. Pull mode makes it hard to prevent malicious attacker stealing data. So push mode is much better since it's the responsibility of the party to decide whether to send data to others.

Reviewer: ```suggestion
- Data transmission across parties should be in push mode instead of pull mode. It's hard to prevent malicious attacker stealing data in pull mode. Push mode is much better since it's the responsibility of the party to decide whether to send data to others.
```

- Tasks should be driven in multi-controller mode. That means tasks should be driven by themselves(who are inside this party) instead of others.

> ### Should this change be within ray or outside?

Reviewer: ```suggestion
```

### Should this change be within ray or outside?

> No

Reviewer: ```suggestion
No, there is no need to change ray repo, the steps should be:
```

![image](https://user-images.githubusercontent.com/19473861/206469580-468ea258-407e-40cf-b059-06a63a25c168.png)

Every Ray cluster will create the Ray tasks specified to its party, meanwhile, it will send the output data  to peer if the downstream Ray task are not specified to its own party. For example, in ALICE cluster, `mean.party('alice').remote(w_a, w_b)` will create a Ray task in ALICE cluster, but it wouldn't be created in BOB cluster. Due to `w_b` is an output data from BOB cluster, therefore, in ALICE cluster, it will be inserted a `recv_op` barrier to `mean` task, and in BOB cluster, it will be inserted a `send_op` barrier as a downstream of `w_b = my_bob.get_weights.remote()` task.

### Benefits 
Compared with running the DAG in one Ray cluster, we have the following significant benefits:
- It's very close to Ray native programming pattern. But the DAG could be ran across Ray clusters.
- Very restricted and clear data perimeters. we only transmit the data which is needed by others. And all of the data couldn't be fetched in any way if we don't allow.

> - If we run the DAG in one Ray cluster, data transmition is in PULL-BASED mode. For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.

Reviewer: ```suggestion
- If we run the DAG in one Ray cluster, data transmission is in PULL-BASED mode. For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.
```

> - If we run the DAG in one Ray cluster, data transmition is in PULL-BASED mode. For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.

Reviewer: ```suggestion
- If we run the DAG in one shared Ray cluster, and the data transmission is in PULL-BASED mode, For example, if BOB invokes `ray.get(f.options("PARTY_ALICE").remote())`, the return object is pulled by BOB, as a result, ALICE don't have the knowledge for that. In this proposal, it's in PUSH-BASED mode. ALICE has the knowledge for that there is a data object will be sent to BOB. This is a significant advantage of multi-controller mode.
```

- All the disadvantages are addressed in this proposal: **code distribution**, **data privacy**, **task-attacks**, **illegal invasion**, **deserialization vulnerability**, and etc.
- Brings all Ray advantages to the local Ray cluster, like highly performance RPC, fault tolerance, task scheduling/resource management, and other Ray ecosystem libraries（ Ray datasets, Ray train and Ray serve） for local training.

## Compatibility, Deprecation, and Migration Plan
None.

## Test Plan and Acceptance Criteria
- Test should include unit tests and integration tests(release tests).
- Tests should be enabled in CI.

> - Benchmark on across parties data transmition for typical federated learning case.

Reviewer: ```suggestion
- Benchmark on across parties data transmission for typical federated learning case.
```
